File: src/main/java/pl/allegro/experiments/chi/chiserver/domain/experiments/administration/stop/StopExperimentCommand.java
package pl.allegro.experiments.chi.chiserver.domain.experiments.administration.stop;

import com.google.common.base.Preconditions;
import pl.allegro.experiments.chi.chiserver.domain.experiments.Experiment;
import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentsRepository;
import pl.allegro.experiments.chi.chiserver.domain.experiments.administration.PermissionsAwareExperimentRepository;

public class StopExperimentCommand {
    private final String experimentId;
    private final ExperimentsRepository experimentsRepository;
    private final PermissionsAwareExperimentRepository permissionsAwareExperimentRepository;

    StopExperimentCommand(
            String experimentId,
            ExperimentsRepository experimentsRepository,
            PermissionsAwareExperimentRepository permissionsAwareExperimentRepository) {
        Preconditions.checkNotNull(experimentId);
        Preconditions.checkNotNull(experimentsRepository);
        Preconditions.checkNotNull(permissionsAwareExperimentRepository);
        this.experimentId = experimentId;
        this.experimentsRepository = experimentsRepository;
        this.permissionsAwareExperimentRepository = permissionsAwareExperimentRepository;
    }

    public void execute() {
        Experiment experiment = permissionsAwareExperimentRepository.getExperimentOrException(experimentId);
        validate(experiment);
        Experiment stoppedExperiment = experiment.stop();
        experimentsRepository.save(stoppedExperiment);
    }

    private void validate(Experiment experiment) {
        if (!experiment.isActive()) {
            throw new StopExperimentException("Experiment is not ACTIVE: " + experimentId);
        }
    }
}

File: src/main/kotlin/pl/allegro/experiments/chi/chiserver/domain/experiments/administration/stop/StopExperimentCommand.kt
package pl.allegro.experiments.chi.chiserver.domain.experiments.administration.stop

import pl.allegro.experiments.chi.chiserver.domain.experiments.Experiment
import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentsRepository
import pl.allegro.experiments.chi.chiserver.domain.experiments.administration.PermissionsAwareExperimentGetter

class StopExperimentCommand(
        private val experimentId: String,
        private val experimentsRepository: ExperimentsRepository,
        private val permissionsAwareExperimentGetter: PermissionsAwareExperimentGetter
) {

    fun execute() {
        val experiment = permissionsAwareExperimentGetter.getExperimentOrException(experimentId)
        validate(experiment)
        val stoppedExperiment = experiment.stop()
        experimentsRepository.save(stoppedExperiment)
    }

    private fun validate(experiment: Experiment) {
        if (!experiment.isActive()) {
            throw StopExperimentException("Experiment is not ACTIVE: ${experimentId}")
        }
    }
}